### PR TITLE
sc-network: expose `add_to_priority_group` and `remove_from_priority_…

### PR DESCRIPTION
**Pull request by ordian**
A follow up to https://github.com/paritytech/substrate/pull/3376#issuecomment-520506177.
Needed for https://github.com/paritytech/polkadot/issues/1763.

polkadot companion: https://github.com/paritytech/polkadot/pull/1800

### CODE REVIEW

File: client/network/src/service.rs [hunk 1]
```diff
@@ -1025,6 +1043,27 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 			.to_worker
 			.unbounded_send(ServiceToWorkerMsg::OwnBlockImported(hash, number));
 	}
+
+	// Utility function to extract `PeerId` from each `Multiaddr` for priority group updates.
+	fn parse_multiaddr(&self, peers: HashSet<Multiaddr>) -> Result<Vec<(PeerId, Multiaddr)>, String> {
```
Comment (mxinden) [inline client/network/src/service.rs:1048]:
```suggestion
	fn split_multiaddr_and_peer_id(&self, peers: HashSet<Multiaddr>) -> Result<Vec<(PeerId, Multiaddr)>, String> {
```

I would find this more descriptive. What do you think?

File: client/network/src/service.rs [hunk 2]
```diff
@@ -1025,6 +1043,27 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 			.to_worker
 			.unbounded_send(ServiceToWorkerMsg::OwnBlockImported(hash, number));
 	}
+
+	// Utility function to extract `PeerId` from each `Multiaddr` for priority group updates.
```
Comment (mxinden) [inline client/network/src/service.rs:1047]:
```suggestion
	/// Utility function to extract `PeerId` from each `Multiaddr` for priority group updates.
	///
	/// Returns an `Err` if one of the given addresses is invalid or contains an
	/// invalid peer ID (which includes the local peer ID).
```

Just to make sure this is not used somewhere where the local peer ID is valid.

File: client/network/src/service.rs [hunk 3]
```diff
@@ -1004,6 +988,40 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		Ok(())
 	}
 
+	/// Add peers to a peerset priority group.
+	///
+	/// Each `Multiaddr` must end with a `/p2p/` component containing the `PeerId`.
+	///
+	/// Returns an `Err` if one of the given addresses is invalid or contains an
+	/// invalid peer ID (which includes the local peer ID).
+	pub fn add_to_priority_group(&self, group_id: String, peers: HashSet<Multiaddr>) -> Result<(), String> {
```
Comment (mxinden) [inline client/network/src/service.rs:997]:
```suggestion
	pub async fn add_to_priority_group(&self, group_id: String, peers: HashSet<Multiaddr>) -> Result<(), String> {
```

Declaring this method as `async` would grant us a greater implementation flexibility in the future, e.g. a bounded channel between `NetworkService` and `NetworkWorker`.

@tomaka what do you think of requiring all new methods on `NetworkService` to be able to yield?

@ordian Are you able to `await` this method within Polkadot?

**Review by mxinden — COMMENTED** [review]
Sorry for the late review @ordian.

File: client/network/src/service.rs (hunk 3 above)
Comment (ordian) [reply client/network/src/service.rs:997]: yes, awaiting is fine, but would we need to change all three methods then (set_, add_, remove_)?

Comment (ordian) [reply client/network/src/service.rs]: Done in b1c28b2.

Comment (tomaka) [reply client/network/src/service.rs]: :+1: As eventually it'd be nice if the queue from the service to the worker was bounded as well.

File: client/network/src/service.rs [hunk 4]
```diff
@@ -1004,6 +988,45 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		Ok(())
 	}
 
+	/// Add peers to a peerset priority group.
+	///
+	/// Each `Multiaddr` must end with a `/p2p/` component containing the `PeerId`.
+	///
+	/// Returns an `Err` if one of the given addresses is invalid or contains an
+	/// invalid peer ID (which includes the local peer ID).
+	// NOTE: even though this function is currently sync, it's marked as async for
```
Comment (mxinden) [inline client/network/src/service.rs:997]:
```suggestion
	//
	// NOTE: even though this function is currently sync, it's marked as async for
```

File: client/network/src/service.rs [hunk 5]
```diff
@@ -1004,6 +988,45 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
...
+		Ok(())
+	}
+
+	/// Remove peers from a peerset priority group.
+	///
+	/// Each `Multiaddr` must end with a `/p2p/` component containing the `PeerId`.
+	///
+	/// Returns an `Err` if one of the given addresses is invalid or contains an
+	/// invalid peer ID (which includes the local peer ID).
+	// NOTE: even though this function is currently sync, it's marked as async for
```
Comment (mxinden) [inline client/network/src/service.rs:1019]:
```suggestion
	//
	// NOTE: even though this function is currently sync, it's marked as async for
```

**Review by mxinden — APPROVED** [review]
Looks good to me. Small suggestion. I am fine either way.

**Review by tomaka — APPROVED** [review]

**Comment by ordian** [comment]
bot merge

**Comment by ordian** [comment]
bot merge

**Pull request merged**